src/analyzer/analyzerworker.cpp

```diff
@@ -0,0 +1,305 @@
+#include "analyzer/analyzerworker.h"
+
+#include <typeinfo>
+
+#include <QtDebug>
+#include <QThread>
+
+#ifdef __VAMP__
+#include "analyzer/analyzerbeats.h"
+#include "analyzer/analyzerkey.h"
+#include "analyzer/vamp/vampanalyzer.h"
+#endif
+#include "analyzer/analyzergain.h"
+#include "analyzer/analyzerebur128.h"
+#include "analyzer/analyzerwaveform.h"
+
+#include "sources/soundsourceproxy.h"
+#include "util/compatibility.h"
+#include "util/event.h"
+#include "util/timer.h"
+#include "util/trace.h"
+
+namespace {
+    // Analysis is done in blocks.
+    // We need to use a smaller block size, because on Linux the AnalyzerWorker
+    // can starve the CPU of its resources, resulting in xruns. A block size
+    // of 4096 frames per block seems to do fine.
+    const SINT kAnalysisChannels = mixxx::AudioSource::kChannelCountStereo;
+    const SINT kAnalysisFramesPerBlock = 4096;
+    const SINT kAnalysisSamplesPerBlock =
+        kAnalysisFramesPerBlock * kAnalysisChannels;
+} // anonymous namespace
+
+// Measured in 0.1%,
+// 0 for no progress during finalize
+// 1 to display the text "finalizing"
+// 100 for 10% step after finalize
+// NOTE: If this is changed, change woverview.cpp slotAnalyzerProgress().
+#define FINALIZE_PROMILLE 1
+
+// --- CONSTRUCTOR ---
+AnalyzerWorker::AnalyzerWorker(UserSettingsPointer pConfig, int workerIdx, bool batchJob) :
+    m_pConfig(pConfig),
+    m_analyzelist(),
+    m_batchJob(batchJob),
+    m_workerIdx(workerIdx),
+    m_sampleBuffer(kAnalysisSamplesPerBlock),
+    m_exit(false),
+    m_pauseRequested(false),
+    m_qm(),
+    m_qwait() {
+}
+
+// --- DESTRUCTOR ---
+AnalyzerWorker::~AnalyzerWorker() {
+    qDebug() << "Ending AnalyzerWorker";
+    // free resources
+    m_progressInfo.sema.release();
+ 
+    QListIterator<Analyzer*> it(m_analyzelist);
+    while (it.hasNext()) {
+        Analyzer* an = it.next();
+        delete an;
+    }
+}
+
+void AnalyzerWorker::nextTrack(TrackPointer newTrack) {
+    m_qm.lock();
+#if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
+    m_currentTrack.clear();
+#else
+    m_currentTrack.reset();
+#endif
+    m_currentTrack = newTrack;
+    m_qwait.wakeAll();
+    m_qm.unlock();
+}
+void AnalyzerWorker::pause() {
+    m_pauseRequested = true;
+}
+void AnalyzerWorker::resume() {
+    m_qm.lock();
+    m_qwait.wakeAll();
+    m_qm.unlock();
+}
+
+void AnalyzerWorker::endProcess() {
+    m_exit = true;
```

Comment:
m_exit is set to false in AnalyzerWorker constructor, and set to true just once when the manager calls this method from its thread (the UI thread). The workers never modify this value, only check which value it has. It is different than the case with pause, where the workers modify it.
I can put it inside the mutex, but that won't change anything because the workers don't block on mutexes to check the value.